File: BinaryTree.java

public class BinaryTree {

    // Node class
    static class Node {

        int data;
        Node left;
        Node right;

        // Constructor
        Node(int data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }

    }

    // Root node of the tree
    Node root;

    // Constructor
    public BinaryTree() {
        root = null;
    }

    // Preorder Traversal
    // Root -> Left -> Right
    public void preorder(Node node) {

        if (node == null) {
            return;
        }

        System.out.print(node.data + " ");

        preorder(node.left);

        preorder(node.right);

    }

    // Inorder Traversal
    // Left -> Root -> Right
    public void inorder(Node node) {

        if (node == null) {
            return;
        }

        inorder(node.left);

        System.out.print(node.data + " ");

        inorder(node.right);

    }

    // Postorder Traversal
    // Left -> Right -> Root
    public void postorder(Node node) {

        if (node == null) {
            return;
        }

        postorder(node.left);

        postorder(node.right);

        System.out.print(node.data + " ");

    }

    public static void main(String[] args) {

        // Create Binary Tree
        BinaryTree tree = new BinaryTree();

        /*
                 1
               /   \
              2     3
             / \   / \
            4   5 6   7
        */

        tree.root = new Node(1);

        tree.root.left = new Node(2);
        tree.root.right = new Node(3);

        tree.root.left.left = new Node(4);
        tree.root.left.right = new Node(5);

        tree.root.right.left = new Node(6);
        tree.root.right.right = new Node(7);

        // Traversals
        System.out.print("Preorder: ");
        tree.preorder(tree.root);

        System.out.println();

        System.out.print("Inorder: ");
        tree.inorder(tree.root);

        System.out.println();

        System.out.print("Postorder: ");
        tree.postorder(tree.root);

    }

}

File: BinaryTree.c
#include <stdio.h>
#include <stdlib.h>

// Node Structure
struct Node {
    int data;
    struct Node *left;
    struct Node *right;
};

// Create a New Node
struct Node* createNode(int value) {

    struct Node *newNode = (struct Node*)malloc(sizeof(struct Node));

    newNode->data = value;
    newNode->left = NULL;
    newNode->right = NULL;

    return newNode;
}

// Preorder Traversal
void preorder(struct Node *root) {

    if(root == NULL)
        return;

    printf("%d ", root->data);

    preorder(root->left);
    preorder(root->right);
}

// Inorder Traversal
void inorder(struct Node *root) {

    if(root == NULL)
        return;

    inorder(root->left);

    printf("%d ", root->data);

    inorder(root->right);
}

// Postorder Traversal
void postorder(struct Node *root) {

    if(root == NULL)
        return;

    postorder(root->left);
    postorder(root->right);

    printf("%d ", root->data);
}

int main() {

    /*
              10
            /    \
          20      30
         /  \    /  \
       40   50 60   70
    */

    struct Node *root = createNode(10);

    root->left = createNode(20);
    root->right = createNode(30);

    root->left->left = createNode(40);
    root->left->right = createNode(50);

    root->right->left = createNode(60);
    root->right->right = createNode(70);

    printf("Preorder Traversal:\n");
    preorder(root);

    printf("\n\n");

    printf("Inorder Traversal:\n");
    inorder(root);

    printf("\n\n");

    printf("Postorder Traversal:\n");
    postorder(root);

    printf("\n");

    return 0;
}
